eval.ml

```diff
@@ -199,34 +200,46 @@ let reduce conf =
        trace = v :: conf.trace;
        pc = pc';
      }
-  | Invalidate (e, l, xs) ->
+  | Osr (e, v, l, osr) ->
      let b = get_bool (eval conf e) in
-     if b then
+     if not b then
        { conf with
          pc = pc';
        }
      else begin
-       let add env x =
-         match Env.find x conf.env with
-         | exception Not_found -> raise (Unbound_variable x)
-         | v -> Env.add x v env
+       let add (env, heap) = function
+         | OsrConst (x, e) ->
+           (Env.add x (Const (eval conf e)) env,
+            heap)
+         | OsrMut (x, OsrExp e) ->
+           let a = Address.fresh () in
+           (Env.add x (Mut a) env,
+            Heap.add a (Value (eval conf e)) heap)
+         | OsrMut (x, OsrUndef) ->
+           let a = Address.fresh () in
+           (Env.add x (Mut a) env,
+            Heap.add a Undefined conf.heap)
```

Comment:
I am not raising this for performance reasons.

Manual object pooling like you suggest might of course be possible in some cases but I think letting the hypothetical GC deal with it is just fine. Note: its not always possible. You can create a new value e.g. "[mut a = 3+2]".

I was more concerned for the proof. It's a globally visible effect that our old invalidate did not have.